### PR TITLE
RGB to Gray ConvertColor Preprocessor

### PR DESCRIPTION
**Pull request by vurusovs**
### Details:
 - C and Py APIs added in scope of https://github.com/openvinotoolkit/openvino/pull/15050

### Tickets:
- 79492

### CODE REVIEW

File: src/core/src/preprocess/preprocess_impls.cpp [hunk 1]
```diff
@@ -193,14 +194,6 @@ bool InputInfo::InputInfoImpl::build(const std::shared_ptr<Model>& model,
     if (node.get_partial_shape() != context.model_shape()) {
         need_validate = true;  // Trigger revalidation if input parameter shape is changed
     }
-    // Check final shape
-    OPENVINO_ASSERT(node.get_partial_shape().compatible(context.model_shape()),
```
Comment (ilyachur) [inline src/core/src/preprocess/preprocess_impls.cpp:197]: Can we disable this check only for gray transformation?

Comment (vurusovs) [reply src/core/src/preprocess/preprocess_impls.cpp:197]: It was a mistake from my side, this check is fully right for RGB->Gray conversion too

File: src/bindings/python/tests/test_graph/test_preprocess.py [hunk 2]
```diff
@@ -344,8 +345,12 @@ def test_graph_preprocess_steps(algorithm, color_format1, color_format2, is_fail
         assert "is not convertible to" in str(e.value)
     else:
         function = custom_processor.build()
-        input_data = np.array([[[[1, 2, 3], [4, 5, 6], [7, 8, 9]]]]).astype(np.float32)
-        expected_output = np.array([[[[0, 3, 6], [1, 4, 7], [2, 5, 8]]]]).astype(np.float32)
+        input_data = np.array([[[[1, 2, 3], [4, 5, 6], [7, 8, 9]]], 
+                               [[[1, 2, 3], [4, 5, 6], [7, 8, 9]]], 
+                               [[[1, 2, 3], [4, 5, 6], [7, 8, 9]]]]).astype(np.float32)
+        expected_output = np.array([[[[0, 3, 6], [1, 4, 7], [2, 5, 8]]],
```
Comment (ilyachur) [inline src/bindings/python/tests/test_graph/test_preprocess.py:351]: Why do we need these changes?

File: src/core/src/preprocess/color_utils.cpp [hunk 3]
```diff
@@ -20,8 +20,10 @@ std::unique_ptr<ColorFormatInfo> ColorFormatInfo::get(ColorFormat format) {
         res.reset(new ColorFormatInfoI420_ThreePlanes(format));
         break;
     case ColorFormat::RGB:
+        res.reset(new ColorFormatRGB(format));
+        break;
     case ColorFormat::BGR:
-        res.reset(new ColorFormatNHWC(format));
+        res.reset(new ColorFormatBGR(format));
```
Comment (ilyachur) [inline src/core/src/preprocess/color_utils.cpp:26]: Why do you change something in RGB/BGR logic?

File: src/core/src/preprocess/color_utils.cpp [hunk 4]
```diff
@@ -20,8 +20,10 @@ std::unique_ptr<ColorFormatInfo> ColorFormatInfo::get(ColorFormat format) {
...
         break;
```
Comment (ilyachur) [inline src/core/src/preprocess/color_utils.cpp:27]: Do you need to extend this switch-case for GRAY?

File: src/tests/ie_test_utils/common_test_utils/CMakeLists.txt [hunk 5]
```diff
@@ -18,6 +18,7 @@ function(add_common_utils ADD_TARGET_NAME)
                 PUBLIC
                     gtest
                     gtest_main
+                    gmock
```
Comment (ilyachur) [inline src/tests/ie_test_utils/common_test_utils/CMakeLists.txt:21]: Why did you add a new dependency?

File: src/tests/ie_test_utils/common_test_utils/test_assertions.hpp [hunk 6]
```diff
@@ -5,6 +5,7 @@
 #pragma once
 
 #include <gtest/gtest.h>
+#include <gmock/gmock-matchers.h>
```
Comment (ilyachur) [inline src/tests/ie_test_utils/common_test_utils/test_assertions.hpp:8]: It can be reverted

Comment (vurusovs) [reply src/tests/ie_test_utils/common_test_utils/test_assertions.hpp:8]: `EXPECT_THAT` (https://github.com/openvinotoolkit/openvino/blob/974650ba6d7a0a17cddcfd944c285e47663f919b/src/tests/ie_test_utils/common_test_utils/test_assertions.hpp#L82) macros is from <gmock/gmock-matchers.h>. So each target which uses `OV_EXPECT_THROW` (based on `EXPECT_THAT`) should be linked with `gmock` by itself.

In this case I decided to add this dependency directly to  `commonTestUtils` target

File: src/tests/ie_test_utils/common_test_utils/CMakeLists.txt (hunk 5 above)
Comment (vurusovs) [reply src/tests/ie_test_utils/common_test_utils/CMakeLists.txt:21]: gmock `EXPECT_THAT` macros is used in https://github.com/openvinotoolkit/openvino/blob/974650ba6d7a0a17cddcfd944c285e47663f919b/src/tests/ie_test_utils/common_test_utils/test_assertions.hpp#L82

File: src/bindings/python/tests/test_graph/test_preprocess.py (hunk 2 above)
Comment (vurusovs) [reply src/bindings/python/tests/test_graph/test_preprocess.py:351]: The test contains a mistake - input tensor with ColorFormat::RGB means "C" channel should be equal to 3. In this PR I introduce strict logic with "C == 3", so the test started to fail

File: src/core/src/preprocess/color_utils.cpp (hunk 3 above)
Comment (vurusovs) [reply src/core/src/preprocess/color_utils.cpp:26]: `ColorFormatBGR` contains specific computation of tensor shape - https://github.com/openvinotoolkit/openvino/pull/13855/files#diff-49bd0f7677452175eb6f598dfb1a537321692e488278c206b8a15d363993ebbcR104

File: src/core/src/preprocess/color_utils.cpp (hunk 4 above)
Comment (vurusovs) [reply src/core/src/preprocess/color_utils.cpp:27]: By default for GRAY case it will be `ColorFormatInfo` class, which default layout is empty (`{}`). If we want to assume GRAY is NHWC input data, so yes, need to extend logic with that case

File: src/tests/ie_test_utils/common_test_utils/test_assertions.hpp (hunk 6 above)
Comment (praasz) [reply src/tests/ie_test_utils/common_test_utils/test_assertions.hpp:8]: if want (and is really required here) include 'gmock' use  `#include <gmock/gmock.h>`  instead, and include `gtest` shouldn't be required as `gmock` depends on it same for linking

File: src/core/tests/preprocess.cpp [hunk 7]
```diff
@@ -209,6 +212,129 @@ TEST(pre_post_process, tensor_element_type_and_scale) {
     EXPECT_EQ(ov::layout::get_layout(f->input(0)), Layout());
 }
 
+class pre_post_process_to_gray : public ::testing::TestWithParam<ColorFormat> {};
```
Comment (praasz) [inline src/core/tests/preprocess.cpp:215]: Detail: Consider rename this class to use `CamelCase` will be more consistent with other class names. e.g `ToGrayTest`
Then final test name in log will look like:
`pre_post_process.ToGrayTest. ....` instead `pre_post_process.pre_post_process_to_gray. ....` there will beno `pre_post_process` duplication.

File: src/core/tests/preprocess.cpp [hunk 8]
```diff
@@ -2,12 +2,15 @@
 // SPDX-License-Identifier: Apache-2.0
 //
 
+#include "common_test_utils/test_assertions.hpp"
+#include "gmock/gmock-matchers.h"
```
Comment (praasz) [inline src/core/tests/preprocess.cpp:6]:
```suggestion
#include "gmock/gmock.h"
```
or remove if included in `test_assertions.hpp`

File: src/core/tests/preprocess.cpp [hunk 9]
```diff
@@ -209,6 +212,129 @@ TEST(pre_post_process, tensor_element_type_and_scale) {
     EXPECT_EQ(ov::layout::get_layout(f->input(0)), Layout());
 }
 
+class pre_post_process_to_gray : public ::testing::TestWithParam<ColorFormat> {};
+
+TEST_P(pre_post_process_to_gray, convert_color_nhwc_to_nhwc) {
+    auto COLOR_FORMAT = GetParam();
```
Comment (praasz) [inline src/core/tests/preprocess.cpp:218]:
```suggestion
    const auto& COLOR_FORMAT = GetParam();
```
To avoid copy, if changed check other similar occurrences.

File: src/core/tests/preprocess.cpp [hunk 10]
```diff
@@ -209,6 +212,129 @@ TEST(pre_post_process, tensor_element_type_and_scale) {
...
+    EXPECT_EQ(f->get_parameters().front()->get_layout(), "NCHW");
+    EXPECT_EQ(f->get_parameters().front()->get_partial_shape(), (PartialShape{Dimension::dynamic(), 3, 2, 2}));
+    EXPECT_EQ(f->get_result()->get_output_partial_shape(0), (PartialShape{Dimension::dynamic(), 2, 2, 1}));
+}
+
+TEST_P(pre_post_process_to_gray, convert_color_asserts) {
+    auto COLOR_FORMAT = GetParam();
+
+    // No "N" dimension name
+    OV_EXPECT_THROW(auto f = create_simple_function(element::f32, PartialShape{Dimension::dynamic(), 2, 2, 1});
```
Comment (praasz) [inline src/core/tests/preprocess.cpp:286]: It looks like these test can be converted to multiple cases and be parameterized where `color format`, `simple model shape`,  and `layout` could be test parameters. Consider use another fixture which use tuple of parameters and make this test simpler 

or maybe refactor each test case to
```cpp
{
    auto f = create_simple_function(element::f32, PartialShape{Dimension::dynamic(), 2, 2, 1});
    auto p = PrePostProcessor(f);
    p.input().tensor().set_layout("DHWC").set_color_format(COLOR_FORMAT);
    p.input().preprocess().convert_color(ColorFormat::GRAY);
    // To point what action throws exception
    OV_EXPECT_THROW(p.build(), ov::AssertFailure, ::testing::HasSubstr("Dimension name 'N' is not found in layout"));
}
```

File: src/core/src/preprocess/color_utils.cpp (hunk 4 above)
Comment (vurusovs) [reply src/core/src/preprocess/color_utils.cpp:27]: Done

File: src/tests/ie_test_utils/common_test_utils/test_assertions.hpp (hunk 6 above)
Comment (vurusovs) [reply src/tests/ie_test_utils/common_test_utils/test_assertions.hpp]: Include full gmock.h for easier extension later

File: src/core/tests/preprocess.cpp (hunk 7 above)
Comment (vurusovs) [reply src/core/tests/preprocess.cpp]: Renamed, but I used snake_case for compatibility with other names

File: src/core/tests/preprocess.cpp (hunk 8 above)
Comment (vurusovs) [reply src/core/tests/preprocess.cpp]: Removed since it's presented in test_assertions.hpp

File: src/core/tests/preprocess.cpp (hunk 9 above)
Comment (vurusovs) [reply src/core/tests/preprocess.cpp]: Done, thanks!

File: src/tests/ie_test_utils/common_test_utils/test_assertions.hpp (hunk 6 above)
Comment (vurusovs) [reply src/tests/ie_test_utils/common_test_utils/test_assertions.hpp:8]: Reverted - gmock.h also introduces `Assign` class which conflicts with our `Assign` op, so it requires additional changes.
Anyway, I would prefer to leave as is in order to not include huger header because purpose is obscure

**Review by praasz — APPROVED** [review]

File: src/core/include/openvino/core/preprocess/color_format.hpp [hunk 11]
```diff
@@ -18,6 +18,7 @@ enum class ColorFormat {
     I420_THREE_PLANES,
     RGB,
     BGR,
+    GRAY,
```
Comment (ilya-lavrenov) [inline src/core/include/openvino/core/preprocess/color_format.hpp:21]: please, add documentation for enum value (s)

File: src/core/src/preprocess/preprocess_steps_impl.cpp [hunk 12]
```diff
@@ -428,6 +428,63 @@ void PreStepsList::add_convert_color_impl(const ColorFormat& dst_format) {
...
+                }
+                if (elem_type.is_integral_number()) {
+                    // Compute in floats due weights are floats
+                    node = std::make_shared<op::v0::Convert>(node, element::f32);
+                    is_converted = true;
+                }
+
+                auto weights_data = context.color_format() == ColorFormat::RGB
+                                        ? std::vector<float>{0.299f, 0.587f, 0.114f}
+                                        : std::vector<float>{0.114f, 0.587f, 0.299f};
```
Comment (ilya-lavrenov) [inline src/core/src/preprocess/preprocess_steps_impl.cpp:461]: can we add a reference where we took these numbers?

Comment (vurusovs) [reply src/core/src/preprocess/preprocess_steps_impl.cpp:462]: Done, thanks!

File: src/core/include/openvino/core/preprocess/color_format.hpp (hunk 11 above)
Comment (vurusovs) [reply src/core/include/openvino/core/preprocess/color_format.hpp]: Added documentation to be handled by doxygen

**Review by ilyachur — APPROVED** [review]
